fs/bcachefs/checksum.c

```diff
@@ -16,53 +20,115 @@
 #include <crypto/skcipher.h>
 #include <keys/user-type.h>
 
-static u64 bch2_checksum_init(unsigned type)
+/*
+ * bch2_checksum state is an abstraction of the checksum state calculated over different pages.
+ * it features page merging without having the checksum algorithm lose its state.
+ * for native checksum aglorithms (like crc), a default seed value will do, and those can easily picked up where they left off.
+ * for hash-like algorithms, a state needs to be stored
+ */
+
+struct bch2_checksum_state {
+	union {
+		u64 seed;
+#ifdef CONFIG_BCACHEFS_ADDITIONAL_HASHES
+		struct xxh64_state h64state;
+#endif
+	};
+	unsigned type;
+};
+
+static void bch2_checksum_init(struct bch2_checksum_state* state)
```

Comment:
AFAIK every use of `bch2_checksum_init` requires `type` to be set beforehand. Would it make sense to pass the state and the type to `bch2_checksum_init` and set the state type in init?

Reply:
I could do that, it's only a matter of preference, no problem